### ISSUE TITLE
CategoricalCrossentropy Label Smoothing Bug for tensors with dim > 2

### ISSUE THREAD

**Issue by osushkov** [tensorflow/tensorflow#39329]
The flag "label_smoothing" in tf.keras.losses.CategoricalCrossentropy does not behave as expected for tensors with dimension > 2. When this parameter is set to non zero value it can on occasion cause severe learning instability and divergence. Looking at the source code the current code is as follows (found [here](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/losses.py#L1521)):

```
def _smooth_labels():
    num_classes = math_ops.cast(array_ops.shape(y_true)[1], y_pred.dtype)
    return y_true * (1.0 - label_smoothing) + (label_smoothing / num_classes)
```
When the y_true tensor has dimension > 2 (for example when using a U-net architecture to classify each pixel of an image), the num_classes will be incorrect.
I think the correct way to look up the num_classes is to get the size of dim=-1:

`num_classes = math_ops.cast(array_ops.shape(y_true)[-1], y_pred.dtype)`

**Comment by amahendrakar** [comment]
@osushkov,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here and also the TensorFlow version you are using. Thanks!

**Comment by osushkov** [comment]
The TF version is as in the linked code in my bug description, 2.2.0. I do not have a code snippet at hand to reproduce this as I came upon it working on a larger project with many dependencies. The error in the code is fairly obvious, as the linked function will only be valid for tensors of rank 2, whereas CategoricalCrossentropy is expected to work for any rank.

**Comment by pavithrasv** [comment]
This has been fixed in https://github.com/tensorflow/tensorflow/commit/5f1ee72e97e09da9171b2c226d4f10f05e89a38a. Thank you!
